### ISSUE TITLE
A bug in the Normalisation neuron

### ISSUE THREAD

**Issue by jzstark** [owlbarn/owl#210]
On the forum @royson mentioned an `Exception: Owl_exception.NOT_BROADCASTABLE` error caused by using `normalisation` in a neural network. Here is his script to reproduce it:

```ocaml
#require "owl_top";;
open Owl;;
open Neural.S;;
open Neural.S.Graph;;
open Algodiff.S;;

let make_network input_shape =
  input input_shape
  |> normalisation ~decay:0.9 (* normalisation layer causes the exception below. *)
  |> lambda (fun x -> Maths.(x / F 256.))
  |> conv2d [|5;5;1;32|] [|1;1|] ~act_typ:Activation.Relu
  |> max_pool2d [|2;2|] [|2;2|]
  |> dropout 0.1
  |> fully_connected 1024 ~act_typ:Activation.Relu
  |> linear 10 ~act_typ:Activation.Softmax
  |> get_network in

let network =
  let x, _, y = Dataset.load_mnist_train_data_arr () in
  let network = make_network [|28;28;1|] in
  Graph.print network;
  let params = Params.config
    ~batch:(Batch.Mini 100) ~learning_rate:(Learning_Rate.Adagrad 0.005) 0.01
  in
  Graph.train ~params network x y |> ignore;
  network in

let imgs, _, labels = Dataset.load_mnist_test_data () in
let m = Dense.Matrix.S.row_num imgs in
let imgs = Dense.Ndarray.S.reshape imgs [|m;28;28;1|] in
Graph.model network imgs;; (* <-- Exception occur here *)
``` 

After some investigation, I find out that it is caused by [this line](https://github.com/owlbarn/owl/blob/master/src/base/neural/owl_neural_neuron.ml#L2049) in the source code:
```ocaml
let mu' = Maths.(a * (sum ~axis:l.axis x))
``` 

Here we suppose input `x` has shape `[|a; b; c; d|]`, and `axis=3`, so the shape of `mu'` will be `[|a; b; c; 1|]`, while actually the expected shape of it should be `[|1; 1; 1; d|]` (see [doc](https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization) from Tensorflow).   The shape mismatch therefore leads to that broadcast-related error. 

As comparison, here is a brief description of how `mu` and `var` parameters are computed during training in Keras:
```python
## Users specify axis and pass input data into
## keras.layers.BatchNormalization
axis = 3
input_shape = [10;28;28;5]

## Compute a list of axes to be reduced 
reduction_axes = list(range(len(input_shape)))
del reduction_axes[self.axis] 
# therefore, reduction_axes = [0,1,2]

## Call backend for computing mean and var with `norm_batch_in_training`
_, mean, variance = K.normalize_batch_in_training(
            inputs, reduction_axes, ...)

## In the TF backend, `norm_batch_in_training` calls `moments` function
mean, var = tf.nn.moments(x, reduction_axes)

## `moments` in turn calls `reduce_mean`
mean = math_ops.reduce_mean(y, axes)
var  = math_ops.reduce_mean(
        math_ops.squared_difference(y, array_ops.stop_gradient(mean)),
        axes, ...)
```
The detailed code can be traced in the following order: [BatchNormalization](https://github.com/keras-team/keras/blob/master/keras/layers/normalization.py#L179) -> [normalize_batch_in_training](https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py#L1716) -> [tf.nn.moments](https://github.com/tensorflow/tensorflow/blob/r1.7/tensorflow/python/ops/nn_impl.py#L690) -> [reduce_mean](https://www.tensorflow.org/api_docs/python/tf/reduce_mean).

We can see the core operation here is to use `reduce_mean` to compute the mean of elements across *multiple dimensions* of a ndarray. A similar operation is **[reduce_sum](https://www.tensorflow.org/api_docs/python/tf/reduce_sum)** is also included in Tensorflow. 
To achieve the same effect, I think `sum_slices` in Ndarray module might be of help.
